binary_transparency/firmware/cmd/ft_monitor/impl/ft_monitor.go

```diff
@@ -89,60 +103,77 @@ func Main(ctx context.Context, opts MonitorOpts) error {
 		case entry = <-ec:
 		}
 
-		stmt := entry.Value
-		if stmt.Type != api.FirmwareMetadataType {
-			// Only analyze firmware statements in the monitor.
-			continue
-		}
+		processEntry(entry, c, opts, matcher)
 
-		// Parse the firmware metadata:
-		var meta api.FirmwareMetadata
-		if err := json.Unmarshal(stmt.Statement, &meta); err != nil {
-			glog.Warningf("Unable to decode FW Metadata from Statement %q", err)
-			continue
+		if entry.Index == entry.Root.TreeSize-1 {
+			// If we have processed all leaves in the current checkpoint, then persist this checkpoint
+			// so that we don't repeat work on startup.
+			bs, err := json.Marshal(entry.Root)
+			if err != nil {
+				return fmt.Errorf("failed to marshal checkpoint: %w", err)
+			}
+			ioutil.WriteFile(opts.StateFile, bs, 0o755)
+			glog.Infof("Persisted state: %v", entry.Root)
 		}
+	}
+}
 
-		glog.Infof("Found firmware (@%d): %s", entry.Index, meta)
+func processEntry(entry client.LogEntry, c client.ReadonlyClient, opts MonitorOpts, matcher *regexp.Regexp) {
```

Comment:
is it not better to return an Error to the caller if processEntry fails?

Reply:
Sure, we could take that option. I was going to do that but then lines {131, 140, 146} all looked really suspicious. As that was the older code than the lines in here which already returned errors, I gave them the right of way.

Are you saying you'd prefer all return paths in here that currently warn to return an error?

Reply:
For now we can leave this comment tagged with the new issue then and revisit again!
Perhaps, when we revisit this function to classify errors, then we will have anyway need to change the template of this function to return an error. For now, you can continue the style as now we agreed to treat all of them as Warnings, with an issue open!

Reply:
This is the same conversation as above, so resolving this and we'll track up there :-)